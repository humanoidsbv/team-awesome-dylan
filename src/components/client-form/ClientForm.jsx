import React, { useState, useRef } from 'react';
import PropTypes from 'prop-types';

import styles from './ClientForm.module.css';

function ClientForm({ toggleClientForm, isClientFormVisible, createClient }) {
  const [address, setAddress] = useState('');
  const [clientName, setClientName] = useState('');
  const [branch, setBranch] = useState('');
  const [locality, setLocality] = useState('');
  const [postalCode, setPostalCode] = useState('');
  const [placeholderOne, setPlaceholderOne] = useState('');
  const [placeholderTwo, setPlaceholderTwo] = useState('');
  const [validity, setValidity] = useState({});

  const formRef = useRef(null);

  const handleBlur = event => {
    setValidity({
      ...validity,
      form: formRef.current.checkValidity(),
      [event.target.name]: event.target.checkValidity()
    });
  };

  const resetFormStates = () => {
    setAddress('');
    setBranch('');
    setClientName('');
    setLocality('');
    setPlaceholderOne('');
    setPlaceholderTwo('');
    setPostalCode('');
  };

  const handleCancel = event => {
    event.preventDefault();
    resetFormStates();
    toggleClientForm();
  };

  const handleSubmit = event => {
    event.preventDefault();
    resetFormStates();

    createClient({
      clientName,
      branch,
      address,
      id: Math.random(),
      locality,
      postalCode,
      placeholderOne,
      placeholderTwo
    });
  };

  return (
    <div
      className={`${styles.clientForm} ${!isClientFormVisible &&
        styles.clientFormHide}`}
    >
      <div className={styles.headerClientForm}>
        <span className={styles.headerClientFormText}>Add new client</span>
        <button
          className={styles.cancelButton}
          onClick={handleCancel}
          type="button"
        >
          Cancel
        </button>
        <button
          className={styles.saveButton}
          disabled={validity.form}
          onClick={handleSubmit}
          type="button"
        >
          Save
        </button>
      </div>
      <div className={styles.personalDetailsTab}>Client details</div>
      <form className={styles.form} onSubmit={handleSubmit} ref={formRef}>
        <div className={styles.formColumnOne}>
          <label className={styles.labelName} htmlFor="clientName">
            Client
            <input
              className={`${styles.inputName} ${validity.clientName === false &&
                styles.inputNameInvalid}`}
              maxLength="35"
              minLength="2"
              name="clientName"
              onBlur={handleBlur}
              onChange={({ target }) => setClientName(target.value)}
              required
              value={clientName}
            />
          </label>
          <label className={styles.labelBranch} htmlFor="branch">
            Branch
            <input
              className={`${styles.inputBranch} ${validity.branch === false &&
                styles.inputBranchInvalid}`}
              maxLength="35"
              minLength="2"
              name="branch"
              onBlur={handleBlur}
              onChange={({ target }) => setBranch(target.value)}
              required
              value={branch}
            />
          </label>
          <label className={styles.labelAddress} htmlFor="address">
            Address
            <input
              className={`${styles.inputAddress} ${validity.address === false &&
                styles.inputAddressInvalid}`}
              maxLength="35"
              minLength="2"
              name="address"
              onBlur={handleBlur}
              onChange={({ target }) => setAddress(target.value)}
              required
              value={address}
            />
          </label>
        </div>
        <div className={styles.formColumnTwo}>
          <div className={styles.localityPostalCodeInputs}>
            <label className={styles.labelPostalCode} htmlFor="postalCode">
              Postal code
              <input
                className={`${styles.inputPostalCode} ${validity.postalCode ===
                  false && styles.inputPostalCodeInvalid}`}
                maxLength="35"
                minLength="2"
                name="postalCode"
                onBlur={handleBlur}
                onChange={({ target }) => setPostalCode(target.value)}
                required
                value={postalCode}
              />
            </label>
            <label className={styles.labelLocality} htmlFor="locality">
              City
              <input
                className={`${styles.inputLocality} ${validity.locality ===
                  false && styles.inputLocalityInvalid}`}
                maxLength="35"
                minLength="2"
                name="locality"
                onBlur={handleBlur}
                onChange={({ target }) => setLocality(target.value)}
                required
                value={locality}
              />
            </label>
          </div>
          <label
            className={styles.labelPlaceholderOne}
            htmlFor="placeholderOne"
          >
            Placeholder 1
            <input
              className={`${
                styles.inputPlaceholderOne
              } ${validity.placeholderOne === false &&
                styles.inputPlaceholderOneInvalid}`}
              maxLength="35"
              minLength="2"
              name="placeholderOne"
              onBlur={handleBlur}
              onChange={({ target }) => setPlaceholderOne(target.value)}
              required
              value={placeholderOne}
            />
          </label>
          <label
            className={styles.labelPlaceholderTwo}
            htmlFor="placeholderTwo"
          >
            Placeholder 2
            <input
              className={`${
                styles.inputPlaceholderTwo
              } ${validity.placeholderTwo === false &&
                styles.inputPlaceholderTwoInvalid}`}
              maxLength="35"
              minLength="2"
              name="placeholderTwo"
              onBlur={handleBlur}
              onChange={({ target }) => setPlaceholderTwo(target.value)}
              required
              value={placeholderTwo}
            />
          </label>
        </div>
      </form>
    </div>
  );
}

ClientForm.propTypes = {
  createClient: PropTypes.func.isRequired,
  isClientFormVisible: PropTypes.bool.isRequired,
  toggleClientForm: PropTypes.func.isRequired
};

export default ClientForm;
