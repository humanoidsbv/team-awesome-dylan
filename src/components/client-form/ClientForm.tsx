import React, { useState, useRef } from 'react';

import Input from '../../shared/components/input/Input';
import styles from './ClientForm.module.css';
import { ValidityState } from '../../shared/types';

interface ClientFormProps {
  createClient: (newClient: {}) => void;
  isClientFormVisible: boolean;
  toggleClientForm: () => void;
}

function ClientForm(
  { toggleClientForm, isClientFormVisible, createClient }: ClientFormProps
): React.ReactElement {
  const [address, setAddress] = useState('');
  const [clientName, setClientName] = useState('');
  const [branch, setBranch] = useState('');
  const [locality, setLocality] = useState('');
  const [postalCode, setPostalCode] = useState('');
  const [contactPerson, setContactPerson] = useState('');
  const [contactPhoneNumber, setContactPhoneNumber] = useState('');
  const [validity, setValidity] = useState<ValidityState>({});

  const formRef = useRef(null);

  const handleBlur = (event: React.ChangeEvent<HTMLInputElement>): void => {
    setValidity({
      ...validity,
      form: formRef.current.checkValidity(),
      [event.target.name]: event.target.checkValidity()
    });
  };

  const resetFormStates = (): void => {
    setAddress('');
    setBranch('');
    setClientName('');
    setLocality('');
    setContactPhoneNumber('');
    setContactPerson('');
    setPostalCode('');
  };

  const handleCancel = (): void => {
    resetFormStates();
    toggleClientForm();
  };

  const handleSubmit = (event: React.FormEvent<HTMLFormElement | HTMLButtonElement>): void => {
    event.preventDefault();
    resetFormStates();

    createClient({
      clientName,
      branch,
      address,
      id: Math.random(),
      locality,
      postalCode,
      contactPerson,
      contactPhoneNumber
    });
  };

  return (
    <div
      className={`${styles.clientForm} ${!isClientFormVisible
        && styles.clientFormHide}`}
    >
      <div className={styles.headerClientForm}>
        <span className={styles.headerClientFormText}>Add new client</span>
        <button
          className={styles.cancelButton}
          onClick={handleCancel}
          type="button"
        >
          Cancel
        </button>
        <button
          className={styles.saveButton}
          disabled={validity.form !== true}
          onClick={handleSubmit}
          type="button"
        >
          Save
        </button>
      </div>
      <div className={styles.personalDetailsTab}>Client details</div>
      <form className={styles.form} onSubmit={handleSubmit} ref={formRef}>
        <div className={styles.formColumnOne}>
          <label className={styles.labelName} htmlFor="clientName">
            Client
            <Input
              isValid={validity.clientName !== false}
              maxLength={35}
              minLength={2}
              name="clientName"
              id="clientName"
              onBlur={handleBlur}
              onChange={({target}: React.ChangeEvent<HTMLInputElement>): void => setClientName(target.value)}
              required
              value={clientName}
            />
          </label>
          <label className={styles.labelBranch} htmlFor="branch">
            Branch
            <Input
              isValid={validity.branch !== false}
              maxLength={35}
              minLength={2}
              name="branch"
              onBlur={handleBlur}
              onChange={({target}: React.ChangeEvent<HTMLInputElement>): void => setBranch(target.value)}
              required
              value={branch}
            />
          </label>
          <label className={styles.labelAddress} htmlFor="address">
            Address
            <Input
              isValid={validity.address !== false}
              maxLength={35}
              minLength={2}
              name="address"
              onBlur={handleBlur}
              onChange={({target}: React.ChangeEvent<HTMLInputElement>): void => setAddress(target.value)}
              required
              value={address}
            />
          </label>
        </div>
        <div className={styles.formColumnTwo}>
          <div className={styles.localityPostalCodeInputs}>
            <label className={styles.labelPostalCode} htmlFor="postalCode">
              Postal code
              <Input
                isValid={validity.postalCode !== false}
                maxLength={35}
                minLength={2}
                name="postalCode"
                onBlur={handleBlur}
                onChange={({target}: React.ChangeEvent<HTMLInputElement>): void => setPostalCode(target.value)}
                required
                value={postalCode}
              />
            </label>
            <label className={styles.labelLocality} htmlFor="locality">
              Locality
              <Input
                isValid={validity.locality !== false}
                maxLength={35}
                minLength={2}
                name="locality"
                onBlur={handleBlur}
                onChange={({target}: React.ChangeEvent<HTMLInputElement>): void => setLocality(target.value)}
                required
                value={locality}
              />
            </label>
          </div>
          <label className={styles.labelContactPerson} htmlFor="contactPerson">
            Contact person
            <Input
              isValid={validity.contactPerson !== false}
              maxLength={35}
              minLength={2}
              name="contactPerson"
              onBlur={handleBlur}
              onChange={({target}: React.ChangeEvent<HTMLInputElement>): void => setContactPerson(target.value)}
              required
              value={contactPerson}
            />
          </label>
          <label
            className={styles.labelContactPhoneNumber}
            htmlFor="contactPhoneNumber"
          >
            Contact phone number
            <Input
              isValid={validity.contactPhoneNumber !== false}
              maxLength={35}
              minLength={2}
              name="contactPhoneNumber"
              onBlur={handleBlur}
              onChange={({target}: React.ChangeEvent<HTMLInputElement>): void => setContactPhoneNumber(target.value)}
              required
              value={contactPhoneNumber}
            />
          </label>
        </div>
      </form>
    </div>
  );
}

export default ClientForm;
