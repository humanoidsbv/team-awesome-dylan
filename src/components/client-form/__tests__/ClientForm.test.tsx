import * as React from 'react';
import { shallow } from 'enzyme';

import ClientForm from '../ClientForm';

it('matches snapshot', () => {
  const clientForm = shallow(
    <ClientForm
      isClientFormVisible={true}
      createClient={({}) => {}}
      toggleClientForm={() => null}
    />
  );
  expect(clientForm).toMatchSnapshot();
});

it('does not display the ClientForm by default', () => {
  const toggleClientFormSpy = jest.fn();
  const clientForm = shallow(
    <ClientForm
      isClientFormVisible={true}
      createClient={({}) => {}}
      toggleClientForm={toggleClientFormSpy}
    />
  );
  expect(toggleClientFormSpy).not.toHaveBeenCalled();
});
