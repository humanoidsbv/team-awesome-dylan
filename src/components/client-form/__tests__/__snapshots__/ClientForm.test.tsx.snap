// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<div
  className="clientForm false"
>
  <div
    className="headerClientForm"
  >
    <span
      className="headerClientFormText"
    >
      Add new client
    </span>
    <button
      className="cancelButton"
      onClick={[Function]}
      type="button"
    >
      Cancel
    </button>
    <button
      className="saveButton"
      disabled={true}
      onClick={[Function]}
      type="button"
    >
      Save
    </button>
  </div>
  <div
    className="personalDetailsTab"
  >
    Client details
  </div>
  <form
    className="form"
    onSubmit={[Function]}
  >
    <div
      className="formColumnOne"
    >
      <label
        className="labelName"
        htmlFor="clientName"
      >
        Client
        <styled.input
          id="clientName"
          isValid={true}
          maxLength={35}
          minLength={2}
          name="clientName"
          onBlur={[Function]}
          onChange={[Function]}
          required={true}
          value=""
        />
      </label>
      <label
        className="labelBranch"
        htmlFor="branch"
      >
        Branch
        <styled.input
          isValid={true}
          maxLength={35}
          minLength={2}
          name="branch"
          onBlur={[Function]}
          onChange={[Function]}
          required={true}
          value=""
        />
      </label>
      <label
        className="labelAddress"
        htmlFor="address"
      >
        Address
        <styled.input
          isValid={true}
          maxLength={35}
          minLength={2}
          name="address"
          onBlur={[Function]}
          onChange={[Function]}
          required={true}
          value=""
        />
      </label>
    </div>
    <div
      className="formColumnTwo"
    >
      <div
        className="localityPostalCodeInputs"
      >
        <label
          className="labelPostalCode"
          htmlFor="postalCode"
        >
          Postal code
          <styled.input
            isValid={true}
            maxLength={35}
            minLength={2}
            name="postalCode"
            onBlur={[Function]}
            onChange={[Function]}
            required={true}
            value=""
          />
        </label>
        <label
          className="labelLocality"
          htmlFor="locality"
        >
          Locality
          <styled.input
            isValid={true}
            maxLength={35}
            minLength={2}
            name="locality"
            onBlur={[Function]}
            onChange={[Function]}
            required={true}
            value=""
          />
        </label>
      </div>
      <label
        className="labelContactPerson"
        htmlFor="contactPerson"
      >
        Contact person
        <styled.input
          isValid={true}
          maxLength={35}
          minLength={2}
          name="contactPerson"
          onBlur={[Function]}
          onChange={[Function]}
          required={true}
          value=""
        />
      </label>
      <label
        className="labelContactPhoneNumber"
        htmlFor="contactPhoneNumber"
      >
        Contact phone number
        <styled.input
          isValid={true}
          maxLength={35}
          minLength={2}
          name="contactPhoneNumber"
          onBlur={[Function]}
          onChange={[Function]}
          required={true}
          value=""
        />
      </label>
    </div>
  </form>
</div>
`;
